### PR TITLE
feat(datafusion): basic FileOpener impl

### PR DESCRIPTION
**Pull request by tshauck**
This PR does a very basic implementation of the `FileOpener` trait in datafusion by a new `ZarrFileOpener`. In a follow up PR, I'll use the non-async zarr reader for the local file case as well as fill out other options (like projection).

Please let me know if you have any feedback.

Thanks!

### CODE REVIEW

**Comment by tshauck** [comment]
It also seems like the action isn't running 😢

File: src/datafusion/file_opener.rs
Comment (tshauck) [inline src/datafusion/file_opener.rs:85]: I'm a little worried about how `lat_lon_example.zarr` actually being a directory will cause challenges downstream. E.g. `object_store.head(&path)` will throw an error, which is why this is manually constructed.

Comment (maximedion2) [reply src/datafusion/file_opener.rs:82]: Oh you mean that the datafusion concept of a file meta isn't well adapted for a directory? Does the Opener struct have to work with a file meta (the one defined in datafusion)? I guess we should keep this as close as possible to other datafusion file opener implementations, but then, the fact that Zarr doesn't exactly expose a single file that can be atomically consumed (e.g. different variables are stored across different files) means it's pretty different from a parquet or csv... In any case, sounds like a slightly more manual way of constructing the file meta is a fine work around for now?

**Comment by maximedion2** [comment]
Sounds good. I took a quick look, I'll review more in depth tomorrow or Friday, are you ready to merge this? I'm okay with merging small PRs relatively quickly, and incrementally getting to where we want, or slightly bigger PRs with more stuff (e.g. going straight to something with projections and filters), at least at this stage.

I'm still looking into improving the stream for the async reader, it's getting tricky, especially with applying predicates (where I read some of the arrays, then decide if I need to read the other requested arrays, for a given chunk, that's making it hard to interleave I/O and decompression). I'll see what I can come up with in the next few days.

File: src/datafusion/file_opener.rs
Comment (tshauck) [reply src/datafusion/file_opener.rs:82]: Yeah, it doesn't really have a concept of treating a directory as having different parts for a datasets. It does work well with lists of files (like a bunch of parquet files), but it mostly lists them, passes them to the `FileOpener::open` method, and coalesces them. So e.g., not sure how this would work: https://docs.rs/datafusion/latest/datafusion/datasource/listing/struct.ListingTable.html

I think things are fine for now (constructing manually), but we'll see which issues we hit and if we can work around them hear, or maybe opportunities to improve DataFusion upstream.

**Comment by tshauck** [comment]
Cool, yeah, lmk what you think when you have a closer look, though generally ok with where this sits... To your point, I think keeping things small to medium, especially to start is good. Things'll inevitably get more heterogenous as we go.

File: src/datafusion/file_opener.rs
Comment (maximedion2) [reply src/datafusion/file_opener.rs:82]: Right right, and I read the code too fast, I see now that you're implementing a trait from `datafusion`, `FileOpener`, which exposes a method that takes in a file meta, so of course "the Opener struct has to work with the file meta", the trait requires it. Okay let's just go ahead with this for now.

**Review by maximedion2 — APPROVED** [review]

**Comment by maximedion2** [comment]
Okay approved, you can go ahead and merge if you want.

**Comment by tshauck** [comment]
Cool, I don't think I can merge however. Could you give me permissions to merge or do it yourself? The former might be nice in that I could investigate the CI issues.

**Comment by maximedion2** [comment]
> Cool, I don't think I can merge however. Could you give me permissions to merge or do it yourself? The former might be nice in that I could investigate the CI issues.

Sent you an invite with write access! I added some protection rules for main, in particular I added a requirement for 1 approval, but I didn't make it particularly strict, I think as it stands you can self approve, which I think is fine if you're like changing settings for actions, to test, for example. Other than that let's have PRs that we approve before merging (you are already doing that, I'll start doing the same too).

**Comment by tshauck** [comment]
That all sounds good to me! Merging...

**Comment by tshauck** [comment]
One other -- low priority -- thing, might to discuss the default merge strategy? Do you have any strong feelings here? Mine aren't strong, but squash and merge is nice because multiple commits into a branch during review end up as a single one, associated with the PR, the main's history.

I think this is also datafusion proper's strategy... https://github.com/apache/arrow-datafusion/blob/47f4b5a67ac3b327764cbd4c0f42da7ac44854e5/dev/release/README.md?plain=1#L38

**Comment by maximedion2** [comment]
Hmm, I think it is the default? But yeah, that's how I prefer it as well, I do like having a commit that represents a standalone, self contained piece of work that leaves the branch in a working state, so I do usually go for squash and merge.

**Comment by maximedion2** [comment]
Actually, that's interesting, https://stackoverflow.com/questions/49203040/is-there-a-way-to-specify-a-default-option-for-merging-a-pull-request-in-github. it seems the last merge determines what shows up as the default for the next one... in any case, I disabled everything except squash and merge.

**Comment by tshauck** [comment]
Sounds good -- thanks!

**Pull request merged**
